### PR TITLE
add typescript types

### PR DESCRIPTION
**Pull request by vkarpov15**

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
+export default class Kareem {
```
Comment (hasezoey) [inline index.d.ts:1]: i dont know if it would be necessary, but maybe add a `declare module 'Kareem'`

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  static skipWrappedFunction(): SkipWrappedFunction;
+  static overwriteMiddlewareResult(): OverwriteMiddlewareResult;
+
+  pre(name: string | RegExp, fn: Function): this;
+  pre(name: string | RegExp, options: Record<string, any>, fn: Function, error?: any, unshift?: boolean): this;
+  post(name: string | RegExp, fn: Function): this;
+  post(name: string | RegExp, options: Record<string, any>, fn: Function, unshift?: boolean): this;
+
+  clone(): this;
+  merge(other: Kareem, clone?: boolean): Kareem;
```
Comment (hasezoey) [inline index.d.ts:11]:
```suggestion
  merge(other: Kareem, clone?: boolean): this;
```

i dont know the semantics, does merge merge both into a *new* thing or in the the existing one?

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
+export default class Kareem {
+  static skipWrappedFunction(): SkipWrappedFunction;
+  static overwriteMiddlewareResult(): OverwriteMiddlewareResult;
+
+  pre(name: string | RegExp, fn: Function): this;
+  pre(name: string | RegExp, options: Record<string, any>, fn: Function, error?: any, unshift?: boolean): this;
+  post(name: string | RegExp, fn: Function): this;
+  post(name: string | RegExp, options: Record<string, any>, fn: Function, unshift?: boolean): this;
+
+  clone(): this;
```
Comment (hasezoey) [inline index.d.ts:10]:
```suggestion
  clone(): Kareem;
```

the semantics suggest this returns a new separate instance, right?
(to my knowledge `this == Class name` (both are equal for typescript), but it should make suggestions (type hints) more obvious as `this` has the understanding of "existing instance" and "class name" as "separate (new) instance"; at least to me)

**Review by hasezoey — COMMENTED** [review]
Looks good initially, maybe consider adding some comments explaining some of the less obvious types / parameters, like `unshift`, `args` or functions

File: index.d.ts (hunk 2 above)
Comment (vkarpov15) [reply index.d.ts:11]: Should be `this`, `merge` merges `other` into the existing one.

File: index.d.ts (hunk 3 above)
Comment (vkarpov15) [reply index.d.ts:10]: Correct, will fix

File: index.d.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  merge(other: Kareem, clone?: boolean): Kareem;
+
+  createWrapper(name: string, fn: Function, context?: any, options?: any): Function;
+  createWrapperSync(name: string, fn: Function): Function;
+  hasHooks(name: string): boolean;
+  filter(fn: Function): Kareem;
```
[the remark this answers is not in the document]
Comment (vkarpov15) [reply index.d.ts]: No, `filter()` returns a copy, it doesn't modify in-place

File: index.d.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+  wrap(name: string, fn: Function, context: any, args: any[], options?: any): Function;
```
[the remark this answers is not in the document]
Comment (vkarpov15) [reply index.d.ts]: Fixed in 20bb652

File: index.d.ts (hunk 1 above)
Comment (vkarpov15) [reply index.d.ts]: Fixed in 3bfa755

**Review by hasezoey — APPROVED** [review]
Looks better now, it would be even better if there would be some tsdoc (as there already exists jsdoc), but i guess this could wait.
also, please know that i am not that familiar with kareem's api or internals.

**Pull request merged**